### PR TITLE
Allow getSymbolsByUDA if type has private members.

### PR DESCRIPTION
**Pull request by rcorre**
Previously, getSymbolsByUDA would fail if given a type that had _any_
private members (even if they weren't marked with an attribute).
This patch prevents getSymbolsByUDA from checking inaccesible members.
This means private symbols with the desired UDA will not be included.
Resolves #15335.

I'm actually not sure if this is the way to go. On one hand, its nice to not totally fail just because some unrelated member is private. On the other hand, having private members silently excluded might be more suprising than a compilation failure.

Thoughts?

### CODE REVIEW

**Comment by rcorre** [comment]
Oh, and I wasn't sure where to put a test for this. It can't go in `std.traits`, because it the symbol would have to be inaccesible from that module.

**Comment by rcorre** [comment]
Another approach:

```d
/**
 * Gets the names of members of `symbol` that have the given user-defined attribute.
 * This is not recursive; it will not search for symbols within symbols such as
 * nested structs or unions.
 */
template getMembersByUDA(alias symbol, alias attribute)
{
    import std.meta : Filter;

    // use mixin instead of getMember to avoid compile errors on private members
    enum hasSpecificUDA(string name) =
        mixin("hasUDA!(symbol."~name~", attribute)");

    alias getMembersByUDA =
        Filter!(hasSpecificUDA, __traits(allMembers, symbol));
}
```

This avoids the privacy issue but returns strings instead of symbols (meaning the user would have to use `getMember` on their end).

File: std/traits.d [hunk 1]
```diff
@@ -6733,8 +6734,14 @@ template getSymbolsByUDA(alias symbol, alias attribute)
 
     static enum hasSpecificUDA(alias S) = hasUDA!(S, attribute);
     alias StringToSymbol(alias Name) = Identity!(__traits(getMember, symbol, Name));
+
+    // filter out members that are inacessible to due privacy
```
Comment (AndrejMitrovic) [inline std/traits.d]: `to due` => `due to`

File: std/traits.d [hunk 2]
```diff
@@ -6733,8 +6734,14 @@ template getSymbolsByUDA(alias symbol, alias attribute)
...
+    enum isAccessible(alias Name) =
+        __traits(compiles, __traits(getMember, symbol, Name));
+    alias accesibleSymbols = Filter!(isAccessible, __traits(allMembers, symbol));
+
     alias getSymbolsByUDA = Filter!(hasSpecificUDA, TypeTuple!(symbol,
-        staticMap!(StringToSymbol, __traits(allMembers, symbol))));
+        staticMap!(StringToSymbol, accesibleSymbols)));
```
Comment (AndrejMitrovic) [inline std/traits.d]: `accessible`

**Comment by AndrejMitrovic** [comment]
Some insignificant typos but LGTM!

**Comment by rcorre** [comment]
Twas a bad day for spelling and grammar. Fixed both, thanks!

**Comment by JackStouffer** [comment]
The code looks good, but this needs a test. As you already stated, I'm not sure how to go about that.

**Comment by rcorre** [comment]
Other than stuffing it into a random module? :)
I wonder if we should have separate dedicated test module for things like this, though it breaks the convenience of inline unittests.

Ideally I'd love to have a way to make a unittest pretend to be in another module, but that's a whole different discussion.

**Comment by rcorre** [comment]
Actually, my test could import a std type from another module that has private members

**Comment by rcorre** [comment]
Alternately, we could have a 'testtypes' module that unittests in other modules could import individually.

**Comment by AndrejMitrovic** [comment]
> Alternately, we could have a 'testtypes' module that unittests in other modules could import individually.

Not a bad idea actually. Perhaps put it inside of `std/internal`.

**Comment by rcorre** [comment]
> Not a bad idea actually. Perhaps put it inside of std/internal

Looks like there was a precedent for that: `std/internal/test/dummyrange.d`

**Comment by JackStouffer** [comment]
@rcorre not really the same thing, dummy range is a utility used in modules for testing range based code where as you want to put the test for a function inside another module. However, putting it inside `std/internal/test/` sounds like the best option here.

**Comment by rcorre** [comment]
True, just realized that its actually part of the main (non-unittest) functionality. I guess it means 'test' in the sense of 'test a condition' rather than 'unittest'.

Just had a slightly different idea that may allow returning private members as symbols:

```d
template getMembersByUDA(T, alias attribute) {
  import std.string : format;
  import std.meta : Filter;

  template toSymbols(names...) {
    static if (names.length == 1)
      mixin("alias toSymbols = AliasSeq!(T.%s);".format(names[0]));
    else
      mixin("alias toSymbols = AliasSeq!(toSymbols!(names[1..$]), %s);".format(names[0]));
  }

  enum hasSpecificUDA(string name) = mixin("hasUDA!(T."~name~", attribute)");
  alias membersWithAttribute = Filter!(hasSpecificUDA, __traits(allMembers, T));
  alias getMembersByUDA = toSymbols!(membersWithAttribute);
}
```

Don't have time to test thoroughly right now.

**Comment by rcorre** [comment]
second attempt at #3827 allows `getSymbolsByUDA` to retrieve private member

**Comment by Hackerpilot** [comment]
If #3827 is meant to replace this request, close this one and we'll review the other one.

**Pull request closed without merging**